### PR TITLE
Pcp 1217 aug08

### PR DESCRIPTION
**Pull request by james-atwill-hs**

### CODE REVIEW

File: vaultclient/token_auth.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package vaultclient
+
+import "github.com/hashicorp/vault/api"
```
Comment (tom-hadlaw-hs) [inline vaultclient/token_auth.go]: Should probably be a single import block

File: activity/sidecar.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+				scrubber.RemoveFiles()
+				jww.FATAL.Fatalf("Authentication token could no longer be renewed.")
+			}
+		}
+	}
+
+	for _, awsLease := range leases.Current.AWSCredentialLeases {
+		jww.DEBUG.Printf("AWS credential for %q expires at: %v", awsLease.AWSCredential.OutputPath, awsLease.Expiry)
+
+		//If an AWS credential lease is going to expire, assume all of them are and rewrite all of them
```
Comment (tom-hadlaw-hs) [inline activity/sidecar.go]: nitpick: space after //

File: activity/sidecar.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+	go func() {
+
+		jww.DEBUG.Printf("Performing auto renewal check on startup.")
+
+		renewLeases(ctx, currentConfig, vaultClient)
+
+		renewTicker := time.NewTicker(util.Flags.RenewInterval)
+		defer renewTicker.Stop()
+
+		jobCompletionTicker := time.NewTicker(18 * time.Second)
```
Comment (tom-hadlaw-hs) [inline activity/sidecar.go]: Where does 18 come from, should that be a constant ?

File: activity/init.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	}
+
+	vaultClient := vaultclient.NewVaultClient(serviceAccountToken,
+		calculateSecretPrefix(currentConfig, serviceSecretPrefix),
+		k8sLoginPath,
+		k8sAuthRole)
+
+	err := vaultClient.Authenticate()
+
+	if err != nil {
```
Comment (shelby-moore) [inline activity/init.go]: Could do this all on one line above.

File: activity/sidecar.go [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+	jww.DEBUG.Printf("Will renew all credentials expiring before %v", threshold)
+
+	jww.DEBUG.Printf("Auth token: expires? %v at %v", leases.Current.AuthTokenLease.CanExpire, leases.Current.AuthTokenLease.ExpiresAt)
+	// Does our authentication token expire soon?
+	if leases.Current.AuthTokenLease.CanExpire {
+		if leases.Current.AuthTokenLease.ExpiresAt.Before(threshold) {
+
+			err := vaultClient.RenewSelf(ctx, util.Flags.RenewLeaseDuration)
+
+			if err != nil {
```
Comment (shelby-moore) [inline activity/sidecar.go]: If the next renewal interval is going to happen after the token expires, should we fatal here?

File: activity/sidecar.go [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+		jww.DEBUG.Printf("SSH certificate %q is valid before %v", certificateFilename, validBefore)
+
+		// If the cert expires before "threshold" o'clock, then we should renew it. If renewing
+		// fails, then there are three outcomes. 1) fails because of permission denied, in which case we exit
+		// 2) fails because of another reason, but is still not expired (in which case we log, but keep going), or
+		// 3) fails because of another reason, but the ssh certificate is now invalid, in which case we exit.
+		if validBefore != uint64(ssh.CertTimeInfinity) && validBefore < uint64(threshold.Unix()) {
+			err := sshsigning.SignKey(vaultClient.Delegate, sshLease.OutputPath, sshLease.VaultMount, sshLease.VaultRole)
+
+			if errwrap.Contains(err, "Code: 403") {
```
Comment (shelby-moore) [inline activity/sidecar.go]: Just matching on 403 should be sufficient?

File: activity/sidecar.go (hunk 3 above)
Comment (shelby-moore) [inline activity/sidecar.go]: Why is this hardcoded to 18 seconds?

File: activity/sidecar.go [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+		jww.INFO.Printf("Lease renewal interval %v, completion check 10s", util.Flags.RenewInterval)
+		for {
+			select {
+			case <-ctx.Done():
+				jww.INFO.Printf("stopping renewal")
+				return
+			case <-renewTicker.C:
+				jww.INFO.Printf("Renewal Heartbeat")
+				renewLeases(ctx, currentConfig, vaultClient)
+			case <-checkKubeAPITicks:
```
Comment (shelby-moore) [inline activity/sidecar.go]: If we aren't implementing this we should leave it out for now

File: activity/cleanup.go [hunk 8]
```diff
@@ -0,0 +1,20 @@
+package activity
+
+import (
+	"github.com/hootsuite/vault-ctrl-tool/leases"
+	"github.com/hootsuite/vault-ctrl-tool/scrubber"
+	"github.com/hootsuite/vault-ctrl-tool/util"
+	jww "github.com/spf13/jwalterweatherman"
+)
+
+func PerformCleanup() {
```
Comment (shelby-moore) [inline activity/cleanup.go]: Nothing is using this

File: activity/init.go [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+	"github.com/hootsuite/vault-ctrl-tool/cfg"
+	"github.com/hootsuite/vault-ctrl-tool/kv"
+	"github.com/hootsuite/vault-ctrl-tool/leases"
+	"github.com/hootsuite/vault-ctrl-tool/scrubber"
+	"github.com/hootsuite/vault-ctrl-tool/sshsigning"
+	"github.com/hootsuite/vault-ctrl-tool/vaultclient"
+	jww "github.com/spf13/jwalterweatherman"
+)
+
+func PerformInitTasks(currentConfig cfg.Config, serviceAccountToken, serviceSecretPrefix, k8sLoginPath, k8sAuthRole *string) {
```
Comment (shelby-moore) [inline activity/init.go]: Nothing is using this?

File: activity/sidecar.go [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt)
+	signal.Notify(c, syscall.SIGTERM)
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				jww.INFO.Printf("stopping renewal")
```
Comment (shelby-moore) [inline activity/sidecar.go]: This log is misleading if we're not actively doing anything in the sidecar.

File: activity/sidecar.go (hunk 7 above)
Comment (shelby-moore) [reply activity/sidecar.go]: If we are going to implement we should try and use constants for statuses. The k8s sdk probably has them already.

**Comment by shelby-moore** [comment]
These changes need unit tests, which I think will uncover needed structural changes in the code.

**Pull request merged**